lib/modem_slm/modem_slm.c

```diff
@@ -0,0 +1,502 @@
+		err = uart_err_check(uart_dev);
+		if (err) {
+			uint32_t now = k_uptime_get_32();
+
+			if (now - start_time > UART_ERROR_DELAY_MS) {
+				LOG_ERR("UART check failed: %d", err);
+				return -EIO;
+			}
+			k_sleep(K_MSEC(10));
+		}
+	} while (err);
+
+	/* Register async handling callback */
+	err = uart_callback_set(uart_dev, uart_callback, NULL);
+	if (err) {
+		LOG_ERR("Cannot set callback: %d", err);
+		return -EFAULT;
+	}
+
+	/* Enable TX */
+	k_sem_give(&tx_done);
+	/* Enable RX */
+	err = uart_rx_enable(uart_dev, uart_rx_buf[0], sizeof(uart_rx_buf[0]), UART_RX_TIMEOUT_US);
+	if (err && err != -EBUSY) {
+		LOG_ERR("UART RX failed: %d", err);
+	} else {
+		next_buf = uart_rx_buf[1];
+	}
+
+	return err;
+}
+
+#if (CONFIG_MODEM_SLM_INDICATE_PIN >= 0)
+static struct gpio_callback gpio_cb;
+
+static void gpio_cb_func(const struct device *dev, struct gpio_callback *gpio_cb, uint32_t pins)
+{
+	if ((BIT(CONFIG_MODEM_SLM_INDICATE_PIN) & pins) == 0) {
+		return;
+	}
+
+	if (k_work_delayable_is_pending(&gpio_wakeup_work)) {
+		(void)k_work_cancel_delayable(&gpio_wakeup_work);
+		(void)gpio_pin_set(gpio_dev, CONFIG_MODEM_SLM_WAKEUP_PIN, 0);
+	}
+
+	LOG_INF("Remote indication");
+	if (ind_handler) {
+		ind_handler();
+	}
+}
+#endif  /* CONFIG_MODEM_SLM_INDICATE_PIN */
+
+static int gpio_init(void)
+{
+	int err;
+
+	if (!device_is_ready(gpio_dev)) {
+		LOG_ERR("GPIO controller not ready");
+		return -ENODEV;
+	}
+
+	err = gpio_pin_configure(gpio_dev, CONFIG_MODEM_SLM_WAKEUP_PIN,
+				 GPIO_OUTPUT_INACTIVE | GPIO_ACTIVE_LOW);
+	if (err) {
+		LOG_ERR("GPIO_0 config error: %d", err);
+		return err;
+	}
+
+#if (CONFIG_MODEM_SLM_INDICATE_PIN >= 0)
+	err = gpio_pin_configure(gpio_dev, CONFIG_MODEM_SLM_INDICATE_PIN,
+				 GPIO_INPUT | GPIO_PULL_UP | GPIO_ACTIVE_LOW);
+	if (err) {
+		LOG_ERR("GPIO_0 config error: %d", err);
+		return err;
+	}
+
+	gpio_init_callback(&gpio_cb, gpio_cb_func, BIT(CONFIG_MODEM_SLM_INDICATE_PIN));
+	err = gpio_add_callback(gpio_dev, &gpio_cb);
+	if (err) {
+		LOG_WRN("GPIO_0 add callback error: %d", err);
+	}
+	err = gpio_pin_interrupt_configure(gpio_dev, CONFIG_MODEM_SLM_INDICATE_PIN,
+					   GPIO_INT_LEVEL_LOW);
+	if (err) {
+		LOG_WRN("GPIO_0 enable callback error: %d", err);
+	}
+#endif
+
+	return 0;
+}
+
+int modem_slm_init(slm_data_handler_t handler)
+{
+	int err;
+
+	if (handler != NULL && data_handler != NULL) {
+		LOG_ERR("Already initialized");
+		return -EFAULT;
+	}
+
+	data_handler = handler;
+	ind_handler = NULL;
+	at_state = AT_CMD_OK;
+
+	err = gpio_init();
+	if (err != 0) {
+		LOG_WRN("Failed to init gpio");
+		return err;
+	}
+
+	err = uart_init(uart_dev);
+	if (err) {
+		LOG_ERR("UART could not be initialized: %d", err);
+		return -EFAULT;
+	}
+
+	k_work_init_delayable(&gpio_wakeup_work, gpio_wakeup_wk);
+	k_work_init(&slm_data_work, slm_data_wk);
+	k_work_init_delayable(&uart_recovery_work, uart_recovery_wk);
+
+	return 0;
+}
+
+int modem_slm_uninit(void)
+{
+	uart_rx_disable(uart_dev);
+	k_sleep(K_MSEC(10));
```

Comment:
Could we add a comment to explain why we are sleeping here?

Reply:
My exerience was CMSIS register write takes time so we pause a little here.